File: NXMineView.m

#import "NXMineView.h"
#import <appkit/graphics.h>

id fieldsList;
int gameMode;
BOOL isRunning = NO;
BOOL timerOn = NO;
int bombs;
int fw, fh, fc;
int maxspc;
int temp;
DPSTimedEntry tEntry;

id actTextField = nil;

const char *Fname = {".NXMines"};
BOOL showAll = NO;

int dx[] = {-1,  0,  1, -1, 1, -1, 0, 1};
int dy[] = {-1, -1, -1,  0, 0,  1, 1, 1};

const char *numbers[] = {" ", "1", "2", "3", "4", "5", "6", "7", "8"};
const char *iNumbers[] = {
	"brickPushed.tiff",
	"brick1.tiff",
	"brick2.tiff",
	"brick3.tiff",
	"brick4.tiff",
	"brick5.tiff",
	"brick6.tiff",
	"brick7.tiff",
	"brick8.tiff"};

const char s0[128], s1[128], s2[128];
char *sp0 = (char *)&s0[0];
char *sp1 = (char *)&s1[0];
char *sp2 = (char *)&s2[0];

@implementation NXMineView

void timer(DPSTimedEntry tEntry, double now, void *udata)
{
    id tD = (id)udata;

    temp++;
    [tD setIntValue: temp];
    if (temp == 999) {
	DPSRemoveTimedEntry(tEntry);
	timerOn = NO;
    }
    return;
}

- initFrame:(const NXRect *)aRect
{
    [super initFrame:aRect];

    gameMode = BEGINNER;
    isRunning = NO;
    fieldsList = [List new];
    
    return self;
}

- postInit: sender
{
    [self loadScores];

    [name0 setTextDelegate: self];
    [name1 setTextDelegate: self];
    [name2 setTextDelegate: self];
    [self setGameMode: gameMode];
    return self;
}

- (BOOL)acceptsFirstResponder
{
    return YES;
}

- rightMouseDown:(NXEvent *)theEvent
{
    NXPoint aP;
    int x, y, cnt, aTag;
    id obj;

    if (isRunning == NO) return self;

    aP.x = theEvent->location.x;
    aP.y = theEvent->location.y;
    [self convertPoint:&aP fromView:nil];

    x = (aP.x)/18;
    y = (aP.y)/18;
    cnt = x + (y*fw);

    obj = [fieldsList objectAt: cnt];
    aTag = [obj tag];

    if (aTag & CHECKED)
	return self;

    if (aTag & FLAG) {
	[obj setTag: (aTag - FLAG)];
	[obj setIcon: "brick.tiff"];
	[obj setEnabled: YES];
	[bombDisplay setIntValue: ([bombDisplay intValue] + 1)];
    } else {
	[obj setTag: (aTag | FLAG)];
	[obj setIcon: "brickAndFlag.tiff"];
	[obj setEnabled: NO];
	[bombDisplay setIntValue: ([bombDisplay intValue] - 1)];
    }

    [obj display];
//Win !!!
    if ( (maxspc == bombs) & ![bombDisplay intValue]) {
	[self endGame: YES];
    }


    return self;
}

- initFieldsW:(int)w H:(int)h Bombs:(int)Bombs
{
    int i, g;
    id obj;
    NXRect aRect;

    [self setAutodisplay: NO];

    bombs = Bombs;
    fw = w;
    fh = h;
    fc = w*h;

    [[self window] sizeWindow: 16 + (fw*18)
			     : 16 + 64 + (fh*18)];

    [self getFrame: &aRect];
    aRect.size.width = (fw*18);
    aRect.size.height = (fh*18);
    [self setFrame: &aRect];

    [[self window] display];

    for (i=0; i<h ; i++) {
	for (g=0; g<w; g++) {
	    obj = [Button new];

	    NXSetRect(&aRect, (g*18), (i*18), 18, 18);
	    [obj initFrame:&aRect
		title:""
		tag:0
		target: self
		action: @selector(buttonPushed:)
		key: 0x00
		enabled: YES];
	    [obj setType: NX_MOMENTARYCHANGE];
	    [obj setAutodisplay: NO];
	    [obj setIcon: "brick.tiff"];
	    [obj setAltIcon: "brickPushed.tiff"];
	    [obj setBordered: NO];
	    [self addSubview: obj];
	    [fieldsList addObject: obj];
	}
    }


    [self setAutodisplay: YES];
    return self;
}

- setBeginner: sender
{
    [self setGameMode:BEGINNER];
    return self;
}

- setMedium: sender
{
    [self setGameMode:MEDIUM];
    return self;
}

- setExpert: sender
{
    [self setGameMode:EXPERT];
    return self;
}

- setGameMode:(int)mode
{
    int i;

    if (isRunning == YES) {
	isRunning = NO;
    }

    if ([fieldsList count]>0)
	for (i=0; i<[fieldsList count]; i++) {
	    [[fieldsList objectAt: i] removeFromSuperview];

	[fieldsList freeObjects];
	}

    gameMode = mode;

    switch (gameMode) {
	case BEGINNER:
	[self initFieldsW:10 H:10 Bombs: 10];
	[[self window] setTitle: "NXMines - Beginner"];
	break;
	case MEDIUM:
	[self initFieldsW:20 H:20 Bombs: 60];
	[[self window] setTitle: "NXMines - Medium"];
	break;
	case EXPERT:
	[self initFieldsW:20 H:20 Bombs: 100];
	[[self window] setTitle: "NXMines - Expert"];
	break;
	default: 
	[self initFieldsW:10 H:10 Bombs: 10];
	[[self window] setTitle: "NXMines - Beginner"];
	break;
   }

    [self startGame: self];
    [self display];
    NXPing();
    return self;
}

- startGame:sender
{
    int i;
    int rnd;
    id obj;

    [self setAutodisplay: NO];

    if (timerOn == YES) {
	DPSRemoveTimedEntry(tEntry);
	timerOn = NO;
    }

    for (i=0; i<[fieldsList count]; i++) {
	obj = [fieldsList objectAt: i];
	[obj setTag: 0];
	[obj setEnabled: YES];
	[obj setIcon: "brick.tiff"];
   }

    for (i=0; i<bombs; i++) {
	do {
	    rnd = (int)(random() % fc);
	} while ([[fieldsList objectAt: rnd] tag] == 1);

	[[fieldsList objectAt: rnd] setTag: 1];
    }

    [self setAutodisplay: YES];
    isRunning = YES;

    [bombDisplay setIntValue: bombs];
    [timeDisplay setIntValue: 0];
    [scorePanel display];

    actTextField = nil;
    temp = 0;

    [self display];

    maxspc = fc;
    return self;
}

- endGame:(BOOL)win
{
    id tm, nm;

    isRunning = NO;
    if (timerOn == YES) DPSRemoveTimedEntry(tEntry);
    timerOn = NO;

    if (win == YES) {
	switch (gameMode) {
	case BEGINNER:
	    tm = time0;
	    nm = name0;
	    break; 
	case MEDIUM:
	    tm = time1;
	    nm = name1;
	    break;
	case EXPERT:
	    tm = time2;
	    nm = name2;
	    break;
	}	
	if (temp <= [tm intValue]) {
	    NXRunAlertPanel("You Won!!", "... and you made best time!",
	    				"OK", 0, 0);
	    [tm setIntValue: temp];
	    [nm setEditable: YES];
//	    [nm selectAll: self];
	    [scorePanel makeKeyAndOrderFront: self];
	    actTextField = nm;	    
	} else
	    NXRunAlertPanel("You Won!", "Wow, you made it!",
	    				"OK", 0, 0);
    } else
	NXRunAlertPanel("You Blasted!", "Hmmm, you found a lost mine...",
	    				"Try Again", 0, 0);
    return self;
}

- buttonPushed: sender
{
    int index = [fieldsList indexOf: sender];
    int xx = index % fh;
    int yy = index / fh;
    int i, g;
    int aTag = [sender tag];
    
    if (isRunning == NO) {
	[sender display];
	return self;
    }

    if (aTag & FLAG)
	return self;

    if (aTag & CHECKED)
	return self;

    switch ([sender tag]) {
	case 0:
	[self checkAtX: xx Y: yy];

	if (timerOn == NO) {
	    tEntry = DPSAddTimedEntry(INTERVAL, timer,
					(char *)timeDisplay, PRIORITY);
	    timerOn = YES;
	NXPing();
	}

	break;

	case 1:
	for (i=0; i<fh; i++)
	    for (g=0; g<fw; g++) {
		[self checkOneAtX:g Y:i];
	    }

	[sender setIcon: "brickPushedAndFalseBomb.tiff"];
	NXPing();

	[self display];
	[self endGame: NO];	//BANG!!!!
	break;
    }

//Win !!!
    if ( (maxspc == bombs) & ![bombDisplay intValue]) {
	[self endGame: YES];
    }


    return self;
}

-(int) tagOfX:(int)x Y:(int)y
{
    int aTag;

    if (x<0) return -1;
    if (y<0) return -1;
    if (x>=fw) return -1;
    if (y>=fh) return -1;

    aTag = [[fieldsList objectAt:(x+(y*fw))] tag]; 

//    if (aTag & CHECKED)
//	return -1;

//    if (aTag & FLAG)
//	return -1;

    return aTag;
}

- checkAtX:(int)xx Y:(int)yy
{
    int count = 0, c, i;
    id obj;

    if (xx<0) return self;
    if (yy<0) return self;
    if (xx>=fw) return self;
    if (yy>=fh) return self;

    for (i=0; i<8; i++) {
	c=[self tagOfX: (xx+dx[i]) Y: (yy+dy[i])];
	if (c&FLAG) c-=FLAG;
	if (c&CHECKED) c-=CHECKED;
	if (c==1) count++;
    }

    obj = [fieldsList objectAt:(xx+(yy*fw))];

    [obj setTag: ([obj tag] | CHECKED)];
    [obj setIcon: iNumbers[count]];
    [obj setEnabled: NO ];
    [obj display];


    if (!count) {
	for (i=0; i<8; i++) {
	    c=[self tagOfX: (xx+dx[i]) Y: (yy+dy[i])];
	    if (!(c&(CHECKED|FLAG)))
	        if (!c) [self checkAtX: (xx+dx[i]) Y: (yy+dy[i])];
	}
    }

    maxspc--;
    return self;
}

- checkOneAtX:(int)xx Y:(int)yy
{
    int count = 0, c, i;
    id obj;

    if (xx<0) return self;
    if (yy<0) return self;
    if (xx>=fw) return self;
    if (yy>=fh) return self;

    obj = [fieldsList objectAt:(xx+(yy*fw))];

    c=[self tagOfX: xx Y: yy];

    if (c & CHECKED) return self;
    if (c & FLAG) {
	c -= FLAG;
	if (c) [obj setIcon: "brickPushedAndBomb.tiff"];
	else [obj setIcon: "brickPushedAndNoBomb.tiff"];

	return self;
    }

    if (c==1) {
	[obj setIcon: "brickPushedAndBomb.tiff"];

    } else if (showAll == YES) {

	for (i=0; i<8; i++) {
	    c=[self tagOfX: (xx+dx[i]) Y: (yy+dy[i])];
	    c &= ~(FLAG+CHECKED);
	    if (c==1) count++;
	}

	[obj setIcon: iNumbers[count]];
    }

    [obj setEnabled: NO ];

    return self;
}


- loadScores
{
    NXTypedStream *aStream = NXOpenTypedStreamForFile (Fname, NX_READONLY);
    int i0, i1, i2;
    int x0, y0, x1, y1;
    NXRect aRect;

    if (aStream != NULL) {
	NXReadTypes(aStream, "iii***iiiii",
		&i0, &i1, &i2,
		&sp0, &sp1, &sp2,
		&gameMode,
		&x0, &y0, &x1, &y1
		);

//	[[self window] read: aStream];
//	[scorePanel read: aStream];

	NXCloseTypedStream(aStream);

	[name0 setStringValue: sp0];
	[time0 setIntValue: i0];
	[name1 setStringValue: sp1];
	[time1 setIntValue: i1];
	[name2 setStringValue: sp2];
	[time2 setIntValue: i2];

	[[self window] getFrame: &aRect];
	aRect.origin.x = (NXCoord)x0;
	aRect.origin.y = (NXCoord)y0;
	[[self window] placeWindow: &aRect];

	[scorePanel getFrame: &aRect];
	aRect.origin.x = (NXCoord)x1;
	aRect.origin.y = (NXCoord)y1;
	[scorePanel placeWindow: &aRect];

    } else
	[self resetScores: nil];
    
    return self;
}

- saveScores
{
    NXTypedStream *aStream;
    int i0, i1, i2;
    int x0, y0, x1, y1;

    NXRect aRect;

    aStream = NXOpenTypedStreamForFile (Fname, NX_WRITEONLY);

    if (aStream == NULL)
	NXRunAlertPanel("File Error", "Couldn't save High Scores", "Ugh !",
		NULL, NULL);
    else {
	i0 = [time0 intValue];
	i1 = [time1 intValue];
	i2 = [time2 intValue];
	strcpy(sp0, [name0 stringValue]);
	strcpy(sp1, [name1 stringValue]);
	strcpy(sp2, [name2 stringValue]);

	[[self window] getFrame: &aRect];
	x0 = (int)(aRect.origin.x);
	y0 = (int)(aRect.origin.y);
	[scorePanel getFrame: &aRect];
	x1 = (int)(aRect.origin.x);
	y1 = (int)(aRect.origin.y);

	NXWriteTypes(aStream, "iii***iiiii",
			&i0, &i1, &i2,
			&sp0, &sp1, &sp2,
			&gameMode,
			&x0, &y0, &x1, &y1
			);

//	[[self window] write: aStream];
//	[scorePanel write: aStream];

	NXCloseTypedStream(aStream);
    }
    return self;
}

- resetScores: sender
{
    int flag;

    if (sender!=nil) {
	flag = NXRunAlertPanel(	"Reset High Scores",
				"Really ???", "Yes","No",0);
	if (flag == 0) return self;
    }

    [name0 setStringValue: "Choler"];
    [time0 setIntValue: 999];
    [name1 setStringValue: "Sang"];
    [time1 setIntValue: 999];
    [name2 setStringValue: "Melan"];
    [time2 setIntValue: 999];

    return self;
}

- read:(NXStream *)aStream
{
    printf("Trying to read...\n");
    [super read:aStream];
    return self;
}

- write:(NXStream *)aStream
{
    printf("Trying to write...\n");
    [super read:aStream];
    return self;
}

/* text delegate methods */

- textDidEnd:textObject endChar:(unsigned short)whyEnd
{
    [actTextField setEditable: NO];
    [scorePanel display];
    printf("char: 0x%x\n", whyEnd);
    return self;
}

//- textDidChange: textObject
//{
//    NXColor Backgnd	= NXConvertRGBToColor(253, 136, 22);
//    NXColor Highlight	= NXConvertRGBToColor(253, 50, 10);

//    [textObject setBackgroundColor: Backgnd];
//    [textObject setBackgroundTransparent: YES];
//    [textObject setSelColor: Highlight];

//    return self;
//}

@end

File: NXMineView.swift
//
//  NXMineView.swift
//  NXMines
//
//  Created by Sebestyén Gábor on 2014.06.22.
//
//

import Cocoa


typealias Field = (Int, Int)

func +( x : Field, y : Field ) -> Field {
    return ( x.0 + y.0 , x.1 + y.1 )
}


let Neighbors : [Field] = [
    (-1,-1), (0,-1), (1,-1),
    (-1,0),          (1,0),
    (-1,1),  (0,1),  (1,1)
]



class NXMineView : NSView, NSTextDelegate {
    // outlets
    @IBOutlet var startButton : NSButton!

    @IBOutlet var bombDisplay : NSTextField!
    @IBOutlet var timeDisplay : NSTextField!

    @IBOutlet var scorePanel : NSPanel!

    @IBOutlet var name0 : NSTextField!
    @IBOutlet var name1 : NSTextField!
    @IBOutlet var name2 : NSTextField!

    @IBOutlet var time0 : NSTextField!
    @IBOutlet var time1 : NSTextField!
    @IBOutlet var time2 : NSTextField!


    enum GameMode {
        case beginner, medium, expert
    }
    

    static let SETUP = [
        // mode -> title, width, height, n of bombs to hide
        GameMode.beginner : (title:"Beginner", cols:10, rows:10, bombs:10),
        GameMode.medium : (title:"medium", cols:20, rows:20, bombs:60),
        GameMode.expert : (title:"Expert", cols:20, rows:20, bombs:100)
    ]

    struct GameSetup {
        var mode : GameMode
        var fh : UInt
        var fw : UInt
        var bombs: UInt
        
        var fc : UInt {
            get {
                return fh*fw
            }
        }
    }


    // internal instance vars
    var fieldsList : [MineButton] = []
    var game : GameSetup?
    // number of uncovered blocks
    var maxspc : UInt = 0
    
    var isRunning = false

    var	timer : Timer?
    var temp = 0


    var showAll = false

    // override NSResponder's acceptsFirstResponder property
    override var acceptsFirstResponder : Bool {
        get { return true }
    }

    deinit {
        // kill timer
        if let t = timer {
            t.invalidate()
        }
    }
    
    
    // get initialized when spawning from NIB file
    override func awakeFromNib() {
        super.awakeFromNib()
        
        loadScores()

        initGame(GameMode.beginner)
    }

    

    
    // MARK: == Actions ==
    
    @IBAction func setBeginner(_ sender: AnyObject) {
        initGame(GameMode.beginner)
    }

    @IBAction func setMedium(_ sender: AnyObject) {
        initGame(GameMode.medium)
    }

    @IBAction func setExpert(_ sender: AnyObject) {
        initGame(GameMode.expert)
    }

    
    @IBAction func startGame(_ sender:AnyObject?) {
        
        // kill timer
        if let t = timer {
            t.invalidate()
            timer = nil
        }
        
        self.window!.isAutodisplay = false
        
        // reset mine field
        for mine in fieldsList {
            mine.hasBomb = false
            mine.flagged = false
            mine.visited = false
            
            mine.isEnabled = true
            mine.image = NSImage(named: "brick.tiff")
        }
        
        // distribute bombs
        var rnd = 0
        for _ in 0..<game!.bombs {
            repeat {
                rnd = Int( arc4random_uniform( UInt32(game!.fc) ) );
            } while fieldsList[ rnd ].hasBomb;
            
            fieldsList[ rnd ].hasBomb = true
        }

        for bomb in fieldsList {
            doCalcBombs(bomb)
        }
        
        
        self.window!.isAutodisplay = true
        
        maxspc = game!.fc // set max clickable field
        temp = 0; // reset time counter
        
        // update displays
        bombDisplay.integerValue = Int(game!.bombs)
        timeDisplay.integerValue = 0
        
        // let it go
        isRunning = true
        
    }
    
    
    // Left Click Handler
    @IBAction func buttonPushed(_ sender:MineButton) {
        guard isRunning else {
            return
        }
        
        if sender.pushed {
            return
        }
        
        // check clicked field
        if sender.hasBomb {
            // KA-BOOOM
            for bomb in fieldsList {
                bomb.isEnabled = false
                
                bomb.doUpdateImage()
            }
            
            // mark the clicked field as false
            sender.image = NSImage(named: "brickPushedAndFalseBomb.tiff")
            NSBeep()
            
            // lose
            endGame(false)
            
        } else {
            // explore clicked area
            doUncover( sender.pos )
            
            if checkWinState() {
                // Win !!!
                endGame(true)
            } else if timer == .none {
                // start ticking
                timer = Timer.scheduledTimer(
                    timeInterval: 1,
                    target: self,
                    selector: #selector(NXMineView.tick(_:)),
                    userInfo: nil,
                    repeats: true)
                NSBeep()
            }
        }
    }
    
    
    // Right Click Handler
    @IBAction func rightButtonPushed(_ field:MineButton) {
        guard isRunning else {
            return
        }

        if field.visited {
            return
        }
        
        // swap flag
        field.doSwapFlag()
        
        bombDisplay.integerValue += field.flagged ? 1 : -1
        
        
        if checkWinState() {
            // Win !!!
            endGame(true)
        }
    }



    
    
    func initGame(_ mode : GameMode) {
        isRunning = false

        // remove all bomb buttons
        for btn in fieldsList {
            btn.removeFromSuperview()
        }
        fieldsList.removeAll()


        // initialize table
        if let s = NXMineView.SETUP[mode] {
            game = GameSetup(mode: mode, fh: UInt(s.cols), fw: UInt(s.rows), bombs: UInt(s.bombs))
            
            // set window title
            self.window?.title = s.title
            
            // -- //
            self.window?.isAutodisplay = false

            // Resize everything
            if let aRect = self.window?.frame {
                // resize window
                self.window!.setFrame(
                    NSRect(
                        origin: aRect.origin,
                        size: CGSize(width: 16 + CGFloat(game!.fw)*18, height: 16 + 64 + CGFloat(game!.fh)*18)
                    ),
                    display: true
                )

                // resize minefield view
                self.frame = NSRect(
                    origin: self.frame.origin,
                    size: CGSize(width: CGFloat(game!.fw)*18, height: CGFloat(game!.fh)*18)
                )
            }
            
            // enforce manual UI update
            self.window!.display()


            fieldsList = [MineButton]()
            
            
            for j in 0..<Int(game!.fw) {
                for i in 0..<Int(game!.fh) {
                    let obj = MineButton(frame: NSRect(x: i*18, y:j*18, width:18, height: 18))

                    obj.pos = (i,j)
                    
                    // obj.target = self
                    // obj.action = Selector("buttonPushed:")

                    self.addSubview(obj)
                    fieldsList.append(obj)
                }
            }
            
            self.window!.display()

            // re-enable auto UI update
            self.window!.isAutodisplay = true
            
            // --- //
            
            
        }
    }


    // @private
    func endGame(_ win: Bool) {
        // terminate current game session
        isRunning = false
        
        // stop timer
        if let t = timer {
            t.invalidate()
            timer = nil
        }
        
        
        // status = win ?
        if win {
            var field1 : NSTextField
            var field2 : NSTextField
            switch game!.mode {
            case .beginner:
                field1 = time0!
                field2 = name0!
            case .medium:
                field1 = time1!
                field2 = name1!
            case .expert:
                field1 = time2!
                field2 = name2!
            }
            
            // time beat
            if temp < field1.integerValue {
                let panel = NSAlert()
                
                panel.messageText = "You Won!!"
                panel.informativeText = "And you beat the best time!"
                panel.addButton(withTitle: "OK")
                
                panel.runModal()
                
                
                field1.integerValue = temp
                field2.isEditable = true
                
                scorePanel.makeKeyAndOrderFront(self)
            } else {
                let panel = NSAlert()
                
                panel.messageText = "You Won!!"
                panel.informativeText = "Although best time is not beaten"
                panel.addButton(withTitle: "OK")
                
                panel.runModal()
            }
            
        } else {
            // Player failed, do nothings
        }
    }


    // NSTimer callback
    func tick(_ aTimer : Timer) {
        if temp < 1000 {
            temp += 1
            
            timeDisplay!.integerValue = temp
        }
    }


    static let Neighbors : [Field] = [
        (-1,-1), (0,-1), (1,-1),
        (-1,0),          (1,0),
        (-1,1),  (0,1),  (1,1)
    ]


    func doUncover(_ f : Field) {
        if let bomb = fieldAt(f) {
            doUncover(bomb)
        }
    }

    func doUncover(_ field : MineButton) {
        if field.visited || field.flagged {
            return
        }

        // Mark field
        field.isEnabled = false
        field.visited = true

        // set bomb image accordingly
        field.doUpdateImage()

        maxspc -= 1

        // Phase two, visit neighbor fields
        if field.bombsAround == 0 {
            for r in NXMineView.Neighbors {
                if let neighbor = fieldAt( (field.pos.0 + r.0, field.pos.1 + r.1) ) {
                    if !neighbor.hasBomb {
                        doUncover( neighbor )
                    }
                }
            }
        }
    }


    /**
     * Count surrounding bombs
     */
    func doCalcBombs(_ bomb : MineButton) {
        // count surrounding bombs
        var count : UInt = 0

        for r in NXMineView.Neighbors {
            if let b = fieldAt( (bomb.pos.0 + r.0, bomb.pos.1 + r.1) ) {
                if b.hasBomb {
                    count += 1
                }
            }
        }
        bomb.bombsAround = count
    }



    // MARK: Scores
    func loadScores() {
        let ud = UserDefaults.standard
        
        if let v = ud.string(forKey: "name0") {
            name0.stringValue = v
            time0.integerValue = ud.integer(forKey: "time0")
        } else {
            name0.stringValue = "Choler"
            time0.integerValue = 999
        }

    
        if let v = ud.string(forKey: "name1") {
            name1.stringValue = v
            time1.integerValue = ud.integer(forKey: "time1")
        } else {
            name1.stringValue = "Sang"
            time1.integerValue = 999
        }

    
        if let v = ud.string(forKey: "name2") {
            name2.stringValue = v
            time2.integerValue = ud.integer(forKey: "time2")
        } else {
            name2.stringValue = "Melan"
            time2.integerValue = 999
        }
    }


    func saveScores() {
        let ud = UserDefaults.standard

        ud.set(name0!.stringValue, forKey: "name0")
        ud.set(time0!.integerValue, forKey: "time0")

        ud.set(name1!.stringValue, forKey: "name1")
        ud.set(time1!.integerValue, forKey: "time1")

        ud.set(name2!.stringValue, forKey: "name2")
        ud.set(time2!.integerValue, forKey: "time2")
    }


    func resetScores() {
        // TODO: complete function
    }
    
    
    
    /** Find and return field in mine field */
    func fieldAt( _ f : Field ) -> MineButton? {
        if f.0 >= 0 && f.0 < Int(game!.fw) && f.1 >= 0 && f.1 < Int(game!.fh) {
            return fieldsList[f.0+(f.1*Int(game!.fw))]
        }
        return .none
    }




    func checkWinState() -> Bool {
        // Special case
        // (visited fields).size == FIELDS.size - bombs.count => LOSE
        let nCovered = Int(game!.fc) - fieldsList.filter{ $0.visited }.count
        if nCovered == Int(game!.bombs) {
            return true
        }

        if bombDisplay.integerValue == 0 {

            let nBombs : UInt = fieldsList
                .filter { $0.flagged && $0.hasBomb }
                .reduce(0) { (sum,n) in sum+1 }

            if nBombs == game!.bombs {
                return true
            }
        }
        
        return false
    }
    

    // MARK: == NSTextDelegate protocol ==
    func textDidEndEditing(_ notification: Notification) {
        // FIXME: never invoked!
        if let target : NSTextField = notification.object as? NSTextField {
            target.isEditable = false
        }
    }
}
